tests/test_log.py

```diff
@@ -38,13 +39,25 @@ def render_log():
 
 def test_log():
     expected = replace_link_ids(
-        "\x1b[2;36m[TIME]\x1b[0m\x1b[2;36m \x1b[0m                                                           \x1b]8;id=0;foo\x1b\\\x1b[2mtest_log.py\x1b[0m\x1b]8;;\x1b\\\x1b[2m:33\x1b[0m\n\x1b[2;36m      \x1b[0m\x1b[2;36m \x1b[0mHello from \x1b[1m<\x1b[0m\x1b[1;95mconsole\x1b[0m\x1b[39m \x1b[0m\x1b[33mwidth\x1b[0m\x1b[39m=\x1b[0m\x1b[1;36m80\x1b[0m\x1b[39m ColorSystem.TRUECOLOR\x1b[0m\x1b[1m>\x1b[0m !      \x1b]8;id=0;foo\x1b\\\x1b[2mtest_log.py\x1b[0m\x1b]8;;\x1b\\\x1b[2m:34\x1b[0m\n\x1b[2;36m      \x1b[0m\x1b[2;36m \x1b[0m\x1b[1m[\x1b[0m\x1b[1;36m1\x1b[0m, \x1b[1;36m2\x1b[0m, \x1b[1;36m3\x1b[0m\x1b[1m]\x1b[0m                                                  \x1b]8;id=0;foo\x1b\\\x1b[2mtest_log.py\x1b[0m\x1b]8;;\x1b\\\x1b[2m:35\x1b[0m\n       \x1b[34m╭─\x1b[0m\x1b[34m───────────────────── \x1b[0m\x1b[3;34mlocals\x1b[0m\x1b[34m ─────────────────────\x1b[0m\x1b[34m─╮\x1b[0m                   \n       \x1b[34m│\x1b[0m \x1b[3;33mconsole\x1b[0m\x1b[31m =\x1b[0m \x1b[1m<\x1b[0m\x1b[1;95mconsole\x1b[0m\x1b[39m \x1b[0m\x1b[33mwidth\x1b[0m\x1b[39m=\x1b[0m\x1b[1;36m80\x1b[0m\x1b[39m ColorSystem.TRUECOLOR\x1b[0m\x1b[1m>\x1b[0m \x1b[34m│\x1b[0m                   \n       \x1b[34m╰────────────────────────────────────────────────────╯\x1b[0m                   \n"
+        "\x1b[2;36m[TIME]\x1b[0m\x1b[2;36m \x1b[0m                                                           \x1b]8;id=0;foo\x1b\\\x1b[2mtest_log.py\x1b[0m\x1b]8;;\x1b\\\x1b[2m:34\x1b[0m\n\x1b[2;36m      \x1b[0m\x1b[2;36m \x1b[0mHello from \x1b[1m<\x1b[0m\x1b[1;95mconsole\x1b[0m\x1b[39m \x1b[0m\x1b[33mwidth\x1b[0m\x1b[39m=\x1b[0m\x1b[1;36m80\x1b[0m\x1b[39m ColorSystem.TRUECOLOR\x1b[0m\x1b[1m>\x1b[0m !      \x1b]8;id=0;foo\x1b\\\x1b[2mtest_log.py\x1b[0m\x1b]8;;\x1b\\\x1b[2m:35\x1b[0m\n\x1b[2;36m      \x1b[0m\x1b[2;36m \x1b[0m\x1b[1m[\x1b[0m\x1b[1;36m1\x1b[0m, \x1b[1;36m2\x1b[0m, \x1b[1;36m3\x1b[0m\x1b[1m]\x1b[0m                                                  \x1b]8;id=0;foo\x1b\\\x1b[2mtest_log.py\x1b[0m\x1b]8;;\x1b\\\x1b[2m:36\x1b[0m\n       \x1b[34m╭─\x1b[0m\x1b[34m───────────────────── \x1b[0m\x1b[3;34mlocals\x1b[0m\x1b[34m ─────────────────────\x1b[0m\x1b[34m─╮\x1b[0m                   \n       \x1b[34m│\x1b[0m \x1b[3;33mconsole\x1b[0m\x1b[31m =\x1b[0m \x1b[1m<\x1b[0m\x1b[1;95mconsole\x1b[0m\x1b[39m \x1b[0m\x1b[33mwidth\x1b[0m\x1b[39m=\x1b[0m\x1b[1;36m80\x1b[0m\x1b[39m ColorSystem.TRUECOLOR\x1b[0m\x1b[1m>\x1b[0m \x1b[34m│\x1b[0m                   \n       \x1b[34m╰────────────────────────────────────────────────────╯\x1b[0m                   \n"
     )
     rendered = render_log()
     print(repr(rendered))
     assert rendered == expected
 
 
+def test_log_no_currentframe():
+    original_currentframe = inspect.currentframe
+    inspect.currentframe = lambda: None
```

Comment:
Any concrete ideas on how to do this? With a helper that encapsulates the logic to extract call stack information, we would still have the same problem (unless, e.g., we pass in a flag, but I don't quite like this approach).

Reply:
How about a helper like this:

```python
@classmethod
_get_frame_info(def current_frame:Optional[Frame]) -> Tuple[str, int, Dict]:
    ...
```

If it has the frame it will use it, otherwise it will default to the old but slower method.

In the code you can call is as `Console._get_frame_info(inspect.currentframe())` but in a test, you can explicitly pass None.

Reply:
So something like
~~~ python
    @staticmethod
    def _get_frame_info(frame: Optional[FrameType], offset: int) -> Tuple[str, int, Dict[str, Any]]:
        # Ignore the frame of this local helper
        offset += 1

        if frame is not None:
            # Use the faster currentframe where implemented
            while offset and frame:
                frame = frame.f_back
                offset -= 1
            assert frame is not None
            return frame.f_code.co_filename, frame.f_lineno, frame.f_locals
        else:
            # Fallback to the slower stack
            try:
                frame_info = inspect.stack()[offset]
            except IndexError:
                raise RuntimeError("Invalid stack offset")
            return frame_info.filename, frame_info.lineno, frame_info.frame.f_locals
~~~
This of course works, but what I don't like is that you'd end up getting frame info even when you're passing in `None` to it, which means that one needs to know what's happening inside to fully understand what's going on. Perhaps not a big deal if documented, but I'd find it a bit confusing at first I think.

Personally, I'd rather go with something like
~~~ python
    @staticmethod
    def _caller_frame_info(offset: int) -> Tuple[str, int, Dict[str, Any]]:
        """Get caller frame information.

        Args:
            offset (int): the caller offset within the current frame stack.

        Returns:
            Tuple[str, int, Dict[str, Any]]: A tuple containing the file name, the line number and
                the dictionary of local variables associated with the caller frame.

        Raises:
            RuntimeError: If the stack offset is invalid.
        """
        # Ignore the frame of this local helper
        offset += 1

        frame = inspect.currentframe()
        if frame is not None:
            # Use the faster currentframe where implemented
            while offset and frame:
                frame = frame.f_back
                offset -= 1
            assert frame is not None
            return frame.f_code.co_filename, frame.f_lineno, frame.f_locals
        else:
            # Fallback to the slower stack
            try:
                frame_info = inspect.stack()[offset]
            except IndexError:
                raise RuntimeError("Invalid stack offset")
            return frame_info.filename, frame_info.lineno, frame_info.frame.f_locals
~~~
and pay the price of temporarily monkey patching `inspect.currentframe`, given that this approach is not unheard of in Python 🙂 . WDYT?

Reply:
The monkey patching feels like a greater evil, tbh. It's a personal preference for this project, I've avoided it thus-far. 

How about your second function, but add a "currentframe" callable that defaults to "inspect.currentframe". In the test, you can pass a callable that returns None.

```python
def _caller_frame_info(offset: int, currentframe: Callable = inspect.currentframe) -> Tuple:
    frame = currentframe()
```

I think that qualifies as "dependency inversion"

Reply:
OK, I'll make the change, even though this approach is akin to passing in a Boolean flag to switch between two different behaviours.